x/distribution/keeper/validator.go

```diff
@@ -42,30 +42,58 @@ func (k Keeper) incrementValidatorPeriod(ctx sdk.Context, val sdk.Validator) uin
 	}
 
 	// fetch historical rewards for last period
-	historical := k.GetValidatorHistoricalRewards(ctx, val.GetOperator(), rewards.Period-1)
+	historical := k.GetValidatorHistoricalRewards(ctx, val.GetOperator(), rewards.Period-1).PeriodRewards
 
-	// fet new historical rewards
-	k.SetValidatorHistoricalRewards(ctx, val.GetOperator(), rewards.Period, historical.Plus(current))
+	// set new historical rewards with reference count of 1
+	k.SetValidatorHistoricalRewards(ctx, val.GetOperator(), rewards.Period, types.NewValidatorHistoricalRewards(historical.Plus(current), 1))
 
 	// set current rewards, incrementing period by 1
 	k.SetValidatorCurrentRewards(ctx, val.GetOperator(), types.NewValidatorCurrentRewards(sdk.DecCoins{}, rewards.Period+1))
 
 	return rewards.Period
 }
 
+// increment the reference count for a historical rewards vlaue
+func (k Keeper) incrementReferenceCount(ctx sdk.Context, valAddr sdk.ValAddress, period uint64) {
+	historical := k.GetValidatorHistoricalRewards(ctx, valAddr, period)
+	if historical.ReferenceCount > 1 {
+		panic("reference count should never exceed 1")
+	}
+	historical.ReferenceCount++
+	k.SetValidatorHistoricalRewards(ctx, valAddr, period, historical)
+}
+
+// decrement the reference count for a historical rewards value, and delete if zero references remain
+func (k Keeper) decrementReferenceCount(ctx sdk.Context, valAddr sdk.ValAddress, period uint64) {
+	historical := k.GetValidatorHistoricalRewards(ctx, valAddr, period)
+	if historical.ReferenceCount == 0 {
+		panic("cannot set negative reference count")
+	}
+	historical.ReferenceCount--
+	if historical.ReferenceCount == 0 {
+		k.DeleteValidatorHistoricalReward(ctx, valAddr, period)
+	} else {
+		k.SetValidatorHistoricalRewards(ctx, valAddr, period, historical)
+	}
+}
+
 func (k Keeper) updateValidatorSlashFraction(ctx sdk.Context, valAddr sdk.ValAddress, fraction sdk.Dec) {
 	height := uint64(ctx.BlockHeight())
 	currentFraction := sdk.ZeroDec()
-	currentPeriod := k.GetValidatorCurrentRewards(ctx, valAddr).Period
+	endedPeriod := k.GetValidatorCurrentRewards(ctx, valAddr).Period - 1
```

Comment:
In reference to earlier comment question :
> cwgoes [2:27 PM]
I just changed where `1` was subtracted
it used to be 1 after the period ending at the slash, which was misleading
it's now set correctly in storage (see `updateValidatorSlashFraction`)